katdal/categorical.py

```diff
@@ -591,69 +654,92 @@ def sensor_to_categorical(sensor_timestamps, sensor_values, dump_midtimes, dump_
     transform : callable or None, optional
-        Transform sensor values before discarding repeats and applying greed
+        Transform [unwrapped] sensor values before fixing initial value,
+        applying greed and discarding repeats
+    initial_value : object or None, optional
+        Sensor value [transformed, unwrapped] to use for dump = 0 up to first
+        proper event (force first proper event to start at dump = 0 by default)
+    greedy_values : sequence or None, optional
+        List of [transformed, unwrapped] sensor values considered "greedy"
     allow_repeats : {False, True}, optional
-        If False, discard sensor events that do not change (transformed) value
+        If False, discard sensor events that do not change [transformed] value
 
     Returns
     -------
     data : :class:`CategoricalData` object
-        Constructed categorical dataset
+        Constructed categorical dataset [unwraps any wrapped values]
 
     """
     sensor_timestamps = np.atleast_1d(sensor_timestamps)
     sensor_values = np.atleast_1d(sensor_values)
-    dump_midtimes = np.atleast_1d(dump_midtimes)
+    dump_endtimes = dump_midtimes + 0.5 * dump_period
+    num_dumps = len(dump_endtimes)
+    # Insert an extra prior dump to collect sensor values before first dump
+    dump_endtimes = np.r_[dump_endtimes[0] - dump_period, dump_endtimes]
     # Check if sensor values are objects wrapped in ComparableArrayWrappers
     wrapped_values = len(sensor_values) and isinstance(sensor_values[0],
                                                        ComparableArrayWrapper)
-    # Convert sensor event times to dump indices (pick the dump during which each sensor event occurred)
-    # The last event is fixed at one-past-the-last-dump, to indicate the end of the last segment
-    events = np.r_[dump_midtimes.searchsorted(sensor_timestamps - 0.5 * dump_period), len(dump_midtimes)]
-    # Cull any empty segments (i.e. when multiple sensor events occur within a single dump, only keep last one)
-    # This also gets rid of excess events before the first dump and after the last dump
-    non_empty = np.nonzero(np.diff(events) > 0)[0]
-    sensor_values, events = sensor_values[non_empty], np.r_[events[non_empty], len(dump_midtimes)]
-    # Force first dump to have valid sensor value (use initial value or first proper value advanced to start)
-    if events[0] != 0 and initial_value is not None:
-        if wrapped_values:
-            initial_value = ComparableArrayWrapper(initial_value)
-        sensor_values, events = np.r_[[initial_value], sensor_values], np.r_[0, events]
-    events[0] = 0
+    # Convert sensor event times to dump indices by picking the dump during
+    # which each sensor event occurred (more precisely: closest dump centroid).
+    # Last event is fixed at one-past-last-dump to indicate end of last segment.
+    events = np.r_[dump_endtimes.searchsorted(sensor_timestamps) - 1, num_dumps]
```

Comment:
Appending num_dumps is I think both unnecessary (because it gets appended again after slicing out the interesting period) and harmful (because if there are past-the-end values, `events` is no longer sorted yet you're using `searchsorted` on it).